src/Utilities/DoubleDicts.jl

```diff
@@ -0,0 +1,381 @@
+module DoubleDicts
+
+import MathOptInterface
+
+const MOI = MathOptInterface
+const CI{F,S} = MOI.ConstraintIndex{F,S}
+
+abstract type AbstractDoubleDict{V, D<:AbstractDict{Int64, V}, D2<:AbstractDict{Tuple{DataType, DataType}, D}} end
```

Comment:
Isn't it an `AbstractDict` ? I would maybe do `AbstractDoubleDict{K, V, ...}` <: `AbstractDict{K, V}`.
For `D`, I would just do `D <: AbstractDict{Int64}`. Indeed, for instance, you might want `DoubleDict` to contain `F`. In this case you have `D = AbstractDict{Int64, F}` so the type of the dict is different for each `(F, S)`. If you impose it to be `Dict{Int64, MO.AbstractFunction}` then this will convert the inner type-stable dicts to type unstable dicts `Dict{Int64, MO.AbstractFunction}`. This is the case for UniversalFallback for instance which stores functions, sets pairs

Reply:
I don't think it can be an `AbstractDict`, if we add the key type `K` to the parameters it will always be not concrete because the are `CI{F,S}` for multiple `F,S` pairs. By not typing the key type at all we have a special structure that can dispatch properly and well typed for those "unstable" keys.